docs/tutorials/python/sagas/index.md

```diff
@@ -0,0 +1,551 @@
+    return f"Undoing booking of car: {input.book_car_id}"
+
+
+@activity.defn
+async def undo_book_hotel(input: BookVacationInput) -> str:
+    print(f"Undoing booking of hotel: {input.book_hotel_id}")
+    return f"Undoing booking of hotel: {input.book_hotel_id}"
+
+
+@activity.defn
+async def undo_book_flight(input: BookVacationInput) -> str:
+    print(f"Undoing booking of flight: {input.book_flight_id}")
+    return f"Undoing booking of flight: {input.book_flight_id}"
+```
+
+For this example, if the number of attempts is less than the allowed number of attempts or if the booking ID is invalid, the Activity will raise exceptions to simulate failures, then run the corresponding undo action.
+
+By setting up these compensations, you'll ensure that your system can handle failures gracefully.
+Next, you'll focus on defining shared data classes and constants to support your Activities and Workflows.
+
+## Define Shared Data Classes and Constants
+
+Shared data classes and constants are used to pass data between Activities and Workflows.
+Common mistakes include using mutable data types such as lists or dictionaries, which can cause unexpected behavior.
+
+Also, Task Queues are shared resources that can be used by multiple Workflows and Workers.
+
+Create a new file named `shared.py`:
+
+```python
+from dataclasses import dataclass
+
+
+@dataclass
+class BookVacationInput:
+    book_user_id: str
+    book_car_id: str
+    book_hotel_id: str
+    book_flight_id: str
+    attempts: int
+
+
+TASK_QUEUE_NAME = "saga-task-queue"
+```
+
+These classes and constants will be used in both Activities, Workflows, and Workers.
+
+Workflows order and run the execution of Activities.
+With your Activities and shared data classes defined, the next step is to create the Workflow.
+This Workflow coordinates the execution of Activities and handle compensations to maintain consistency in case of failures.
+
+## Create the Workflow Definition
+
+In the context of Temporal Workflows, compensation refers to the actions taken to roll back a transaction if an error occurs.
+Each step in the Workflow has a corresponding compensation step that is executed in reverse order if the Workflow encounters an error.
+
+This ensures that the system is returned to a consistent state, even in the case of partial failures.
+
+Create a new file named `saga_workflows.py`.
+This file will define your Workflow, which is responsible for executing your Activities in the correct order and handling compensation if necessary.
+
+First, import the necessary modules:
+
+```python
+from datetime import timedelta
+from temporalio import workflow
+from temporalio.common import RetryPolicy
+
+with workflow.unsafe.imports_passed_through():
+    from activities import BookVacationInput, book_car, book_flight, book_hotel
+```
+
+Next, create the `BookWorkflow` class and define the compensation actions, as well as the functions that execute your core logic: `book_car`, `book_hotel`, and `book_flight`.
```

Comment:
You do an excellent job of explaining how you track the compensating activities after the code. But I think it's worth explaining the general pattern here. You're going to call each activity and keep track of the compensations you need to run if they fail so that you can execute the appropriate ones at the end. I don't think you need more than that because your description after the code that explains how you do it is excellent.